File: uSkyBlock-Core/src/main/java/us/talabrek/ultimateskyblock/util/TranslationUtil.java
package us.talabrek.ultimateskyblock.util;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.reflect.TypeToken;
import org.bukkit.Bukkit;
import org.bukkit.inventory.ItemStack;
import org.bukkit.plugin.java.JavaPlugin;

import java.io.File;
import java.io.IOException;
import java.lang.reflect.Type;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

/**
 * Utility for translating Minecraft item/block internal names to localized display names.
 * Downloads language files from Mojang's asset servers on first load.
 */
public class TranslationUtil {
    public static final TranslationUtil INSTANCE = new TranslationUtil();

    public Map<String, String> dictionary = new HashMap<>();

    private TranslationUtil() {}

    /**
     * Returns the localized name for an ItemStack's material.
     * Falls back to the lowercase material name if no translation is available.
     */
    public String getItemLocalizedName(ItemStack itemStack) {
        String name = itemStack.getType().name().toLowerCase(Locale.getDefault());
        if (dictionary.containsKey("item.minecraft." + name)) {
            return dictionary.get("item.minecraft." + name);
        }
        if (dictionary.containsKey("block.minecraft." + name)) {
            return dictionary.get("block.minecraft." + name);
        }
        return name;
    }

    /**
     * Loads item language translations from the plugin's i18n/items folder.
     * If the folder does not exist, fetches language assets from Mojang first.
     */
    public boolean loadItemLang(JavaPlugin plugin) {
        File folder = new File(plugin.getDataFolder(), "i18n/items");
        if (!folder.exists()) {
            fetchLanguageAssets(plugin);
        }

        File[] files = folder.listFiles();
        if (files == null) {
            return false;
        }

        for (File file : files) {
            String tr;
            try {
                tr = Files.readString(file.toPath(), StandardCharsets.UTF_8);
            } catch (IOException e) {
                e.printStackTrace();
                plugin.getLogger().severe("TranslationUtil: error in loading " + file.getName());
                return false;
            }

            Type type = new TypeToken<HashMap<String, String>>() {}.getType();
            dictionary = new Gson().fromJson(new JsonParser().parse(tr).getAsJsonObject(), type);
        }
        plugin.getLogger().info("TranslationUtil: Loaded " + dictionary.size() + " item translations. ");
        return true;
    }

    @SuppressWarnings("deprecation")
    private void fetchLanguageAssets(JavaPlugin plugin) {
        try {
            JsonParser parser = new JsonParser();
            String version = Bukkit.getMinecraftVersion();
            String versionManifestJson = readUrl("https://launchermeta.mojang.com/mc/game/version_manifest.json");
            JsonArray versionsManifest = parser.parse(versionManifestJson).getAsJsonObject().getAsJsonArray("versions");

            URL versionUrl = null;
            for (JsonElement element : versionsManifest) {
                JsonObject obj = element.getAsJsonObject();
                if (obj.has("id") && obj.get("id").getAsString().equals(version)) {
                    versionUrl = new URL(obj.get("url").getAsString());
                    break;
                }
            }
            if (versionUrl == null) {
                plugin.getLogger().severe("TranslationUtil: Unable to find version information: " + version);
                throw new RuntimeException();
            }

            plugin.getLogger().info("TranslationUtil: Fetching metadata of " + version + "...");
            JsonObject versionObj = parser.parse(readUrl(versionUrl.toString())).getAsJsonObject();
            String assetsURL = versionObj.getAsJsonObject("assetIndex").get("url").getAsString();
            if (assetsURL == null) {
                plugin.getLogger().severe("TranslationUtil: Unable to get assets information.");
                throw new RuntimeException();
            }

            plugin.getLogger().info("TranslationUtil: Fetching assets information...");
            JsonObject objects = parser.parse(readUrl(assetsURL)).getAsJsonObject().getAsJsonObject("objects");

            File folder = new File(plugin.getDataFolder(), "i18n/items");
            if (!folder.exists()) {
                folder.mkdirs();
            }

            java.util.regex.Pattern pattern = java.util.regex.Pattern.compile("minecraft/lang/([^\\.]+)\\.json");

            for (Map.Entry<String, JsonElement> entry : objects.entrySet()) {
                String key = entry.getKey();
                java.util.regex.Matcher match = pattern.matcher(key);
                if (match.matches() && match.groupCount() == 1) {
                    String lang = match.group(1);

                    if (!lang.equals("zh_cn")) continue;

                    JsonObject langJsonObject = objects.getAsJsonObject("minecraft/lang/" + lang + ".json");
                    if (langJsonObject == null) {
                        plugin.getLogger().severe("TranslationUtil: Failed to fetch: " + lang + ", internal error.");
                        continue;
                    }
                    plugin.getLogger().info("TranslationUtil: Fetching " + lang + "...");
                    String langJsonDownloadUrl = null;
                    try {
                        String hash = langJsonObject.get("hash").getAsString();
                        langJsonDownloadUrl = "https://resources.download.minecraft.net/" + hash.substring(0, 2) + "/" + hash;
                        URL json = new URL(langJsonDownloadUrl);
                        File file = new File(folder, lang + ".json");
                        Files.copy(json.openStream(), file.toPath(), StandardCopyOption.REPLACE_EXISTING);
                    } catch (Exception e) {
                        plugin.getLogger().severe("Cannot load " + lang + " from mojang, url=" + langJsonDownloadUrl);
                        e.printStackTrace();
                        return;
                    }
                }
            }
        } catch (Exception e) {
            plugin.getLogger().warning("TranslationUtil: Error when get item translation files from mojang. Please check your config or add these files manually.");
            plugin.getServer().getPluginManager().disablePlugin(plugin);
        }
    }

    @SuppressWarnings("deprecation")
    private static String readUrl(String urlString) throws IOException {
        URL url = new URL(urlString);
        return new String(url.openStream().readAllBytes(), StandardCharsets.UTF_8);
    }
}

File: uSkyBlock-Core/src/main/java/us/talabrek/ultimateskyblock/util/TranslationUtil.kt
package us.talabrek.ultimateskyblock.util

import com.google.gson.Gson
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.google.gson.reflect.TypeToken
import org.bukkit.Bukkit
import org.bukkit.inventory.ItemStack
import org.bukkit.plugin.java.JavaPlugin
import java.io.File
import java.lang.reflect.Type
import java.net.URL
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.util.*
import java.util.regex.Matcher
import java.util.regex.Pattern

object TranslationUtil {
    var dictionary = HashMap<String, String>()

    fun getItemLocalizedName(itemStack: ItemStack): String {
        val name: String = itemStack.type.name.lowercase(Locale.getDefault())
        if (dictionary.contains("item.minecraft.$name")) {
            return dictionary.get("item.minecraft.$name").toString()
        }
        if (dictionary.contains("block.minecraft.$name")) {
            return dictionary.get("block.minecraft.$name").toString()
        }
        return name
    }

    fun loadItemLang(plugin: JavaPlugin): Boolean {
        val folder: File = File(plugin.getDataFolder(), "i18n/items")
        if (!folder.exists()) {
            fetchLanguageAssets(plugin)
        }

        for (file in folder.listFiles()!!) {
            var tr: String? = null
            try {
                tr = file.readText()
            } catch (e: java.lang.Exception) {
                e.printStackTrace()
                plugin.logger.severe("TranslationUtil: error in loading " + file.name)
                return false
            }

            val type: Type = object: TypeToken<HashMap<String, String>>() {}.type
            dictionary = Gson().fromJson(JsonParser().parse(tr).asJsonObject, type)

        }
        plugin.logger.info("TranslationUtil: Loaded " + dictionary.toList().size + " item translations. ")
        return true
    }

    private fun fetchLanguageAssets(plugin: JavaPlugin) {
        try {
            val parser = JsonParser()
            val version = Bukkit.getBukkitVersion().split("-".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0]
            val versionsManifest: JsonArray = parser.parse(
                URL("https://launchermeta.mojang.com/mc/game/version_manifest.json").readText()
            ).asJsonObject.getAsJsonArray("versions") // like a array with: {"id": "1.21.1", "type": "release", "url": "https://piston-meta.mojang.com/v1/packages/c4c37a8abcbab7fb31031acfb19b9a3a34efb1ec/1.21.1.json", "time": "2025-07-17T06:36:54+00:00", "releaseTime": "2024-08-08T12:24:45+00:00"}
            var versionUrl: URL? = null
            versionsManifest.forEach {
                if (it is JsonObject && it.has("id") && it.get("id").asString.equals(version)) {
                    versionUrl = URL(it.get("url").asString)
                }
            }
            if (versionUrl == null) {
                plugin.logger.severe("TranslationUtil: Unable to find version information: $version")
                throw Exception()
            }
            // versionUrl like: https://piston-meta.mojang.com/v1/packages/c4c37a8abcbab7fb31031acfb19b9a3a34efb1ec/1.21.1.json
            plugin.logger.info("TranslationUtil: Fetching metadata of $version...")
            val assetsURL = parser.parse(versionUrl!!.readText()).asJsonObject.getAsJsonObject("assetIndex").get("url").asString // "assetIndex": {"id": "26", "sha1": "7eb8873392fc365779dbfea6e2c28fca30a6c6cd", "size": 490976, "totalSize": 432035882, "url": "https://piston-meta.mojang.com/v1/packages/7eb8873392fc365779dbfea6e2c28fca30a6c6cd/26.json"}
            if (assetsURL == null) { // https://piston-meta.mojang.com/v1/packages/7eb8873392fc365779dbfea6e2c28fca30a6c6cd/26.json
                plugin.logger.severe("TranslationUtil: Unable to get assets information.")
                throw Exception()
            }
            plugin.logger.info("TranslationUtil: Fetching assets information...") // like array of "minecraft/lang/zh_cn.json": {"hash": "6a829093df0c8f7d3942e26655290334925a9b23", "size": 475421}
            val objects = parser.parse(URL(assetsURL).readText()).asJsonObject.getAsJsonObject("objects")

            val folder = File(plugin.dataFolder, "i18n/items")
            if (!folder.exists()) {
                folder.mkdirs()
            }
            val pattern: Pattern = Pattern.compile("minecraft/lang/([^\\.]+)\\.json") // minecraft/lang/zh_cn.json -> zh_cn

            objects.entrySet().forEach { (key) ->
                val match: Matcher = pattern.matcher(key)
                if (match.matches() && match.groupCount() == 1) {
                    val lang = match.group(1)

                    if (lang != "zh_cn") return@forEach

                    val langJsonObject = objects.getAsJsonObject("minecraft/lang/$lang.json") // minecraft/lang/zh_cn.json
                    if (langJsonObject == null) {
                        plugin.logger.severe("TranslationUtil: Failed to fetch: $lang, internal error.")
                        return@forEach
                    }
                    plugin.logger.info("TranslationUtil: Fetching $lang...")
                    var langJsonDownloadUrl: String? = null
                    try {
                        val hash: String = langJsonObject.get("hash").asString
                        langJsonDownloadUrl = "https://resources.download.minecraft.net/" + hash.substring(0, 2) + "/" + hash // 6a/6a829093df0c8f7d3942e26655290334925a9b23
                        val json = URL(langJsonDownloadUrl)
                        val file = File(folder, "$lang.json")
                        Files.copy(json.openStream(), file.toPath(), StandardCopyOption.REPLACE_EXISTING)
                    } catch (e: Exception) {
                        plugin.logger.severe("Cannot load $lang from mojang, url=$langJsonDownloadUrl")
                        e.printStackTrace()
                        return
                    }
                }
            }
        } catch (e: Exception) {
            plugin.logger.warning("TranslationUtil: Error when get item translation files from mojang. Please check your config or add these files manually.")
            plugin.server.pluginManager.disablePlugin(plugin)
        }
    }
}
